core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java

```diff
@@ -0,0 +1,183 @@
+  public static final double MAX_FILE_SIZE_DEFAULT_RATIO = 1.80d;
+
+  private int minInputFiles;
+  private int minOutputFiles;
+  private long minFileSize;
+  private long maxFileSize;
+  private long targetFileSize;
+  private long maxGroupSize;
+
+  @Override
+  public String name() {
+    return Strategy.BINPACK.name();
+  }
+
+  @Override
+  public Set<String> validOptions() {
+    return ImmutableSet.of(
+        MIN_INPUT_FILES,
+        MIN_OUTPUT_FILES,
+        MIN_FILE_SIZE_BYTES,
+        MAX_FILE_SIZE_BYTES
+    );
+  }
+
+  @Override
+  public RewriteStrategy options(Map<String, String> options) {
+    targetFileSize = PropertyUtil.propertyAsLong(options,
+        RewriteDataFiles.TARGET_FILE_SIZE_BYTES,
+        PropertyUtil.propertyAsLong(
+            table().properties(),
+            TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+            TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT));
+
+    minFileSize = PropertyUtil.propertyAsLong(options,
+        MIN_FILE_SIZE_BYTES,
+        (long) (targetFileSize * MIN_FILE_SIZE_DEFAULT_RATIO));
+
+    maxFileSize = PropertyUtil.propertyAsLong(options,
+        MAX_FILE_SIZE_BYTES,
+        (long) (targetFileSize * MAX_FILE_SIZE_DEFAULT_RATIO));
+
+    maxGroupSize = PropertyUtil.propertyAsLong(options,
+        RewriteDataFiles.MAX_FILE_GROUP_SIZE_BYTES,
+        RewriteDataFiles.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT);
+
+    minOutputFiles = PropertyUtil.propertyAsInt(options,
+        MIN_OUTPUT_FILES,
+        MIN_OUTPUT_FILES_DEFAULT);
+
+    minInputFiles = PropertyUtil.propertyAsInt(options,
+        MIN_INPUT_FILES,
+        MIN_INPUT_FILES_DEFAULT);
+
+    validateOptions();
+    return this;
+  }
+
+  @Override
+  public Iterable<FileScanTask> selectFilesToRewrite(Iterable<FileScanTask> dataFiles) {
+    return FluentIterable.from(dataFiles)
+        .filter(scanTask -> scanTask.length() < minFileSize || scanTask.length() > maxFileSize);
+  }
+
+  @Override
+  public Iterable<List<FileScanTask>> planFileGroups(Iterable<FileScanTask> dataFiles) {
+    ListPacker<FileScanTask> packer = new BinPacking.ListPacker<>(maxGroupSize, 1, false);
+    List<List<FileScanTask>> potentialGroups = packer.pack(dataFiles, FileScanTask::length);
+    return potentialGroups.stream().filter(group ->
+        numOutputFiles(group) >= minOutputFiles && group.size() >= minInputFiles
+    ).collect(Collectors.toList());
+  }
+
+  private long numOutputFiles(List<FileScanTask> group) {
+    long groupSize = group.stream().mapToLong(FileScanTask::length).sum();
+    return (long) Math.ceil((double) groupSize / targetFileSize);
+  }
+
+  private void validateOptions() {
+    Preconditions.checkArgument(minFileSize >= 0,
+        "Cannot set %s to a negative number, %d < 0",
+        MIN_FILE_SIZE_BYTES, minFileSize);
+
+    Preconditions.checkArgument(maxFileSize > minFileSize,
+        "Cannot set %s greater than %s, %d >= %d",
```

Comment:
nit: `greater than` -> `greater than or equal to`?

Reply:
Wait, can they be equal actually?

Reply:
This is the "error" condition which is the inverse of the "valid condition"
Min File is greater than or equal to Max file size

Reply:
I think technically there really isn't any value in allowing minfile == maxfile == targetfile

Reply:
Agree